File: src/ros2_bridge/mappers/sensor_msgs/point_cloud.rs
//! Typed mapper for `sensor_msgs/msg/PointCloud`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn point_cloud_to_bus(
    msg: ros_env::sensor_msgs::msg::PointCloud,
) -> crate::sensor_msgs::msg::v1::PointCloud {
    crate::sensor_msgs::msg::v1::PointCloud {
        header: Some(crate::ros2_bridge::mappers::std_msgs::header::header_to_bus(msg.header)),
        points: msg
            .points
            .into_iter()
            .map(crate::ros2_bridge::mappers::geometry_msgs::point32::point32_to_bus)
            .collect(),
        channels: msg
            .channels
            .into_iter()
            .map(crate::ros2_bridge::mappers::sensor_msgs::channel_float32::channel_float32_to_bus)
            .collect(),
    }
}

pub(crate) fn point_cloud_to_ros(
    bus: crate::sensor_msgs::msg::v1::PointCloud,
) -> ros_env::sensor_msgs::msg::PointCloud {
    ros_env::sensor_msgs::msg::PointCloud {
        header: crate::ros2_bridge::mappers::std_msgs::header::header_to_ros(
            bus.header.unwrap_or_default(),
        ),
        points: bus
            .points
            .into_iter()
            .map(crate::ros2_bridge::mappers::geometry_msgs::point32::point32_to_ros)
            .collect(),
        channels: bus
            .channels
            .into_iter()
            .map(crate::ros2_bridge::mappers::sensor_msgs::channel_float32::channel_float32_to_ros)
            .collect(),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct SensorMsgsPointCloudMapper;

impl TypedTopicMapper for SensorMsgsPointCloudMapper {
    type Ros = ros_env::sensor_msgs::msg::PointCloud;
    type Bus = crate::sensor_msgs::msg::v1::PointCloud;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(point_cloud_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(point_cloud_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/sensor_msgs/point_cloud.py
"""Generated mapper for `sensor_msgs/msg/PointCloud`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.std_msgs.header import header_to_bus, header_to_ros
from robot_bus.ros2_bridge.mappers.geometry_msgs.point32 import point32_to_bus, point32_to_ros
from robot_bus.ros2_bridge.mappers.sensor_msgs.channel_float32 import channel_float32_to_bus, channel_float32_to_ros

def point_cloud_to_bus(msg):
    from robot_bus.sensor_msgs.msg.v1 import PointCloud as BusMsg

    bus = BusMsg()
    bus.header.CopyFrom(header_to_bus(msg.header))
    bus.points.extend([point32_to_bus(x) for x in msg.points])
    bus.channels.extend([channel_float32_to_bus(x) for x in msg.channels])
    return bus


def point_cloud_to_ros(bus):
    from sensor_msgs.msg import PointCloud as RosMsg

    out = RosMsg()
    out.header = header_to_ros(bus.header)
    out.points = [point32_to_ros(x) for x in bus.points]
    out.channels = [channel_float32_to_ros(x) for x in bus.channels]
    return out


class SensorMsgsPointCloudMapper:
    def ros_msg_type(self):
        from sensor_msgs.msg import PointCloud as RosMsg

        return RosMsg

    def ros_to_bus(self, msg) -> bytes:
        return point_cloud_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        from robot_bus.sensor_msgs.msg.v1 import PointCloud as BusMsg

        bus = BusMsg()
        bus.ParseFromString(payload)
        return point_cloud_to_ros(bus)
